style: Move all the fixup code into a StyleAdjuster struct.

emilio opened the pull request:
This will allow reusing it from text styles, which we need for some corner
cases, like text-align: -moz-center and similar stuff.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16608)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/lib.rs, components/style/properties/longhand/box.mako.rs, components/style/style_adjuster.rs, components/style/properties/properties.mako.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style and layout code, but no tests are modified. Please consider adding a test!

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 75cce524f5fe57a17c0f2b2ab5962c887f68599f with merge 9d86f52d902972ca99e6518bcb39ffc46db6db79...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2066)

emilio commented:
@jryans wanted to take this review, since he's been looking at this lately, so...

r? @jryans

jryans commented on components/style/properties/properties.mako.rs:
```diff
@@ -1568,6 +1569,14 @@ pub mod style_structs {
+                                            dpy: lonhands::display::computed_value::T) {
+                    // NB: We don't touch __servo_display_for_hypothetical_box,
+                    // which is the whole point.
+                    //
```
Probably don't need this extra comment line?

emilio commented:
@bors-servo retry

jryans commented on components/style/properties/properties.mako.rs:
```diff
@@ -2339,208 +2363,9 @@ pub fn apply_declarations<'a, F, I>(device: &Device,
-    // https://fullscreen.spec.whatwg.org/#new-stacking-layer
-    // Any position value other than 'absolute' and 'fixed' are
-    // computed to 'absolute' if the element is in a top layer.
-    % if product == "gecko":
```
`adjust_for_top_layer` seems to be missing these conditions, is it correct?

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 0be84aee4c8aaf79140edbffb76c857c3ee9ea1c with merge a22c2fd271fa2838437189ce1f3534e760c45a49...

emilio commented on components/style/style_adjuster.rs:
```diff
@@ -0,0 +1,245 @@
+    ///    computed to 'absolute' if the element is in a top layer.
+    ///
+    fn adjust_for_top_layer(&mut self) {
+        self.style.mutate_box().set_position(position::absolute);
```
whoopsies

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 5812ada47aa7978b7def2847d1e0843beadcde70 with merge 2e03a24c0c545ef3d23dc0d2070193ff3ff0f2c0...

jryans commented:
Okay, looks good to me!

@bors-servo delegate+

bors-servo commented:
:v: @emilio can now approve this pull request

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16597) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/6880)

emilio commented:
I deserved that.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 25d76d29e270f9527141a8fee7704bd988e2f016 with merge bb861ebb351fd8705165b38f89ad105953a60630...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3620)

emilio commented:
@bors-servo try

That failure is due to the display for hypothetical box stuff. I can leave it untouched, but I want to check if the obvious thing works (probably not).

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit e394bb5be0804b0917a3629b46cf7db94a38fe7b with merge 2cb83d4f8f0d37b1ecbb5bd5e58b248e667b673a...

emilio commented:
Gecko is perfectly green: https://treeherder.mozilla.org/#/jobs?repo=try&revision=5e557ace33f57b0c5869a26b28c8a2bd8f2dc887

So if Servo's fine, then this is good to merge, otherwhise I'll need to add a trivial change to preserve the previous behavior in servo (add a bool to set_adjusted_display).

@bors-servo r=jryans

bors-servo commented:
:pushpin: Commit e394bb5 has been approved by `jryans`

<!-- @bors-servo r=jryans e394bb5be0804b0917a3629b46cf7db94a38fe7b -->

bors-servo commented:
:hourglass: Testing commit e394bb5be0804b0917a3629b46cf7db94a38fe7b with merge e68629029a0277e007e39e6bd3a38904ab30c1cc...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3626)

emilio committed:
style: Move all the fixup code into a StyleAdjuster struct.

This will allow reusing it from text styles, which we need for some
corner cases, like text-align: -moz-center and similar stuff.

emilio commented:
@bors-servo r=jryans

There are a few interesting potential passes showing up the brokennes of our current approach too, but this stuff is tricky to fix, and I don't have time for it right now.

bors-servo commented:
:pushpin: Commit ccf15c0 has been approved by `jryans`

<!-- @bors-servo r=jryans ccf15c0a267e1ee6dac67a1392aec1860dbcd294 -->

bors-servo commented:
:hourglass: Testing commit ccf15c0a267e1ee6dac67a1392aec1860dbcd294 with merge 09f0ff74811405eb95200ee78d936b4378de65fb...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/6543), [arm32](http://build.servo.org/builders/arm32/builds/6754), [arm64](http://build.servo.org/builders/arm64/builds/6789), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6889), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3628), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3657), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6892), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6931), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3582), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3651), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2078)
Approved by: jryans
Pushing 09f0ff74811405eb95200ee78d936b4378de65fb to master...
